hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOffloadableOperation.java

```diff
@@ -0,0 +1,315 @@
+ */
+public class EntryOffloadableOperation extends LockAwareOperation implements BackupAwareOperation, MutatingOperation {
+
+    private EntryProcessor entryProcessor;
+
+    private transient boolean readOnly;
+
+    public EntryOffloadableOperation() {
+    }
+
+    public EntryOffloadableOperation(String name, Data dataKey, EntryProcessor entryProcessor) {
+        super(name, dataKey);
+        this.entryProcessor = entryProcessor;
+
+    }
+
+    @Override
+    public void innerBeforeRun() throws Exception {
+        super.innerBeforeRun();
+
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
+        final ManagedContext managedContext = serializationService.getManagedContext();
+        managedContext.initialize(entryProcessor);
+    }
+
+    @Override
+    public void run() {
+        final long now = getNow();
+        SerializationService serializationService = getNodeEngine().getSerializationService();
+        Object oldValue = recordStore.get(dataKey, false);
+
+        if (!(entryProcessor instanceof Offloadable)) {
+            throw new HazelcastException("EntryProcessor is expected to implement Offloadable for this operation");
+        }
+        readOnly = entryProcessor instanceof ReadOnly;
+        if (readOnly && entryProcessor.getBackupProcessor() != null) {
+            throw new HazelcastException("EntryProcessor.getBackupProcessor() should return null if ReadOnly implemented");
+        }
+
+        boolean shouldCloneForOffloading = OBJECT.equals(mapContainer.getMapConfig().getInMemoryFormat());
+        Object value = shouldCloneForOffloading ? serializationService.toData(oldValue) : oldValue;
+        runOnOffloadedThread(value, ((Offloadable) entryProcessor).getExecutorName(), now);
+    }
+
+    private long getNow() {
+        return Clock.currentTimeMillis();
+    }
+
+    @SuppressWarnings({"unchecked", "checkstyle:anoninnerlength"})
+    private void runOnOffloadedThread(final Object oldValue, String executorName, final long now) {
+        final OperationServiceImpl ops = (OperationServiceImpl) getNodeEngine().getOperationService();
+
+        final Data finalDataKey = dataKey;
+        final String finalCaller = getCallerUuid();
+        final long finalThreadId = threadId;
+        final long finalCallId = getCallId();
+
+        ops.onStartAsyncOperation(this);
+        executorName = executorName.equals(Offloadable.DEFAULT_ASYNC_EXECUTOR) ? ASYNC_EXECUTOR : executorName;
+
+        final ObjectNamespace namespace = new DefaultObjectNamespace(MapService.SERVICE_NAME, name);
+
+        if (!readOnly) {
+            recordStore.lock(finalDataKey, finalCaller, finalThreadId, finalCallId, Long.MAX_VALUE);
+        }
+
+        // execute for key
+        getNodeEngine().getExecutionService().execute(executorName, new Runnable() {
+            @Override
+            public void run() {
+                if (readOnly) {
```

Comment:
Better to create two different `Runnable`s for readonly and write cases... They look like completely separate.